src/sentry/workflow_engine/utils/scopedstats.py

```diff
@@ -0,0 +1,179 @@
+        self._data[key][tags_tuple] += amount
+
+    def set(self, key: str, tags: Tags | None = None, value: int | float = 0) -> None:
+        tags_tuple = _generate_tags_key(tags)
+        self._data[key][tags_tuple] = value
+
+    def merge_into(self, target: _StatsCollector) -> None:
+        """Merge this collector's data into another collector."""
+        for key, tags_data in self._data.items():
+            target_key_data = target._data[key]
+            for tags_tuple, amount in tags_data.items():
+                target_key_data[tags_tuple] += amount
+
+
+class Recorder:
+    """Records statistics during context blocks. Use with record() context manager."""
+
+    __slots__ = ("_data", "_has_recorded")
+
+    def __init__(self) -> None:
+        self._data: dict[str, dict[tuple[tuple[str, str | bool], ...], int | float]] = defaultdict(
+            lambda: defaultdict(int)
+        )
+        self._has_recorded = False
+
+    @contextmanager
+    def record(self) -> Generator[None]:
+        """Context manager for recording statistics. Automatically adds total_recording_duration."""
+        collector = _StatsCollector()
+        parent_collector = _current_collector.get()
+        context_token = _current_collector.set(collector)
+
+        start_time = time.perf_counter()
+
+        try:
+            yield
+        finally:
+            end_time = time.perf_counter()
+            recording_duration = end_time - start_time
+
+            collector.set("total_recording_duration", value=recording_duration)
+
+            self._merge_collector(collector)
+            self._has_recorded = True
+
+            _current_collector.reset(context_token)
+
+            if parent_collector is not None:
+                collector.merge_into(parent_collector)
+
+    def _merge_collector(self, collector: _StatsCollector) -> None:
+        for key, tags_data in collector._data.items():
+            final_key_data = self._data[key]
+            for tags_tuple, amount in tags_data.items():
+                final_key_data[tags_tuple] += amount
+
+    def get_result(
+        self, tag_filter: Tags | None = None, *, require_recording: bool = False
+    ) -> dict[str, int | float]:
+        """Get recorded statistics.
+
+        Args:
+            tag_filter: Only include stats with matching tags
+            require_recording: Raise ValueError if no recording has occurred
+        """
+        if require_recording and not self._has_recorded:
+            raise ValueError(
+                "No recording has occurred. Use recorder.record() context manager first."
+            )
+        return _get_filtered_stats(self._data, tag_filter)
+
+
+def incr(key: str, tags: Tags | None = None, amount: int | float = 1) -> None:
+    """Increment a counter if stats are being recorded."""
+    collector = _current_collector.get()
+    if collector:
+        tags_tuple = _generate_tags_key(tags)
+        collector._data[key][tags_tuple] += amount
```

Comment:
should this just call [`collector.increment`](https://github.com/getsentry/sentry/pull/98756/files#diff-d88949dbecf5783a19fc7cdab829053e8deb4047474feba20a3c1a7cc49e99f9R62-R64)?